File: Nov2020/15Nov/HelpThemOut.java
/*
Problem Name: Help them Out
Problem Link: https://www.hackerearth.com/practice/data-structures/arrays/1-d/practice-problems/algorithm/solve-friends-problem/
*/
import java.util.*;
class HelpThemOut {
    public static void main(String args[] ) throws Exception {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt(), g = -1;
        int dec[] = new int[n], div[] = new int[n];
        for(int i=0;i<n;i++){
            g = in.nextInt();
            while(g>0){
                if(g%2 == 0){
                    g/=2;
                    div[i]++;
                }
                else{
                    g--;
                    dec[i]++;
                }
            }
        }
        for(int i=1;i<n;i++){
            dec[i] += dec[i-1];
            div[i] = Math.max(div[i],div[i-1]);
        }
        System.out.println(dec[n-1] + div[n-1]);
    }
}

File: Nov2020/15Nov/HelpThemOut.cpp
/*
Problem Name: Help them Out
Problem Link: https://www.hackerearth.com/practice/data-structures/arrays/1-d/practice-problems/algorithm/solve-friends-problem/
*/

#include<bits/stdc++.h>
using namespace std;
int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0);             
    int N;
    cin>>N;
    int g[N];
    for(int i=0;i<N;++i)
        cin>>g[i];
    int moves=0,odd=0,sum=0,flag=0;
    while(flag!=1){
        odd=0,sum=0;
        for(int i=0;i<N;++i){
            if(g[i]%2!=0){
                g[i]-=1;
                moves++;
                odd=1;
            }
            sum+=g[i];
        }
        if(sum!=0 && odd!=1){
            moves++;
            for(int i=0;i<N;++i){
                g[i]/=2;
            }
        }
        else if(sum==0){
            flag=1;
        }
    }
    cout<<moves;
    return 0;
}
